geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,249 @@
+
+            grad = state["grad"]
+
+            # compute retract and transport in gradient direction
+            new_point, grad_retr = manifold.retr_transp(point, -step_size * grad, grad)
+            with torch.no_grad():
+                point.copy_(new_point)
+
+            state["grad_retr"] = grad_retr
+
+        # recompute loss at new point
+        phi = self.closure()
+
+        # Store new gradients in state
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+
+            state["new_grad"] = grad
+
+        # roll back parameters to before step
+        with torch.no_grad():
+            for point, old_point in zip(self._params, param_copy):
+                point.copy_(old_point)
+
+        self._step_size_dic[step_size] = phi
+        return phi
+
+    def derphi_(self, step_size):
+        """Compute derivative of phi.
+
+        The derivative of phi is given by computing inner
+        product between all tensor gradients at target point and those at source point.
+        The source gradients are transported to the target point, and both gradients are
+        projected.
+        """
+
+        # Call phi_ to compute gradients; Does nothing if phi_ was
+        # already called with this stepsize during this step
+        self.phi_(step_size)
+
+        derphi = 0
+        for point in self._params:
+            state = self.state[point]
+            if "grad" not in state:
+                continue
+
+            # For some reason using the metric for inner product gives wrong results,
+            # therefore us euclidean product instead.
+            derphi += torch.tensordot(
+                -state["new_grad"],
+                state["grad_retr"],
+                dims=len(state["new_grad"].shape),
+            ).item()
+        return derphi
+
+    def init_loss(self):
+        """Compute loss and gradients at start of line search."""
+
+        loss = self.closure()
+        derphi0 = 0
+        grad_norms = []
+        self._step_size_dic = dict()
+
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norms.append(grad.norm().item())
```

Comment:
do you really need Euclidean norm here and not Riemannian one?

Reply:
No, Riemannian is better, thank you for the observation. I will update this.

Reply:
btw, all norm computations are broadcastable and can factorize across product manifold. if you have a product manifold, you may better have more than one grad norm. 

```
# pseudocode
point = Sphere(10, 3)  # 10 3d spheres
inner = point.inner(v1, v2)
inner.shape
(10,)
```

Does having a factorized inner product makes sense in this optimizer? In Adam this affected adaptive term in a good way

Reply:
I just need the squared sum of gradient norms to perform the line search, so this isn't necessary. But this is good to keep in mind if implementing geometric conjugate gradient (which seems useful).